controllers/machine_controller.go

```diff
@@ -183,9 +185,18 @@ func (r *MachineReconciler) reconcileDelete(ctx context.Context, cluster *cluste
 		}
 	} else {
 		klog.Infof("Deleting node %q for machine %q", m.Status.NodeRef.Name, m.Name)
-		if err := r.deleteNode(ctx, cluster, m.Status.NodeRef.Name); err != nil && !apierrors.IsNotFound(err) {
-			klog.Errorf("Error deleting node %q for machine %q: %v", m.Status.NodeRef.Name, m.Name, err)
-			return ctrl.Result{}, err
+
+		var deleteNodeErr error
+		waitErr := wait.PollImmediate(2*time.Second, 10*time.Second, func() (bool, error) {
+			if deleteNodeErr = r.deleteNode(ctx, cluster, m.Status.NodeRef.Name); deleteNodeErr != nil && !apierrors.IsNotFound(deleteNodeErr) {
+				return false, nil
+			}
+			return true, nil
+		})
+		if waitErr != nil {
+			// TODO: remove m.Name after #1203
+			r.Log.Error(deleteNodeErr, "timed out deleting Machine's node, moving on", "node", m.Status.NodeRef.Name, "machine", m.Name)
+			r.recorder.Eventf(m, corev1.EventTypeWarning, "FailedDeleteNode", "error deleting Machine's node: %v", deleteNodeErr)
```

Comment:
I assume the answer is "no" but it's worth asking: the if condition here is a non-nil waitErr. Is there a chance that waitErr is not nil, but deleteNodeErr is nil (since we're including deleteNodeErr in the event)?

Reply:
no, if waitErr is not nil, deleteNodeErr will definitely not-nil
if deleteNodeErr is nil, then conditionFunc will return true and waitErr will be nil